driver-core/src/main/java/com/datastax/driver/core/TokenRange.java

```diff
@@ -224,13 +224,27 @@ public boolean contains(Token token) {
     // isStart handles the case where the token is the start of another range, for example:
     // * ]1,2] contains 2, but it does not contain the start of ]2,3]
     // * ]1,2] does not contain 1, but it contains the start of ]1,3]
-    private boolean contains(Token token, boolean isStart) {
-        boolean isAfterStart = isStart ? token.compareTo(start) >= 0 : token.compareTo(start) > 0;
-        boolean isBeforeEnd = end.equals(factory.minToken()) ||
-                (isStart ? token.compareTo(end) < 0 : token.compareTo(end) <= 0);
-        return isWrappedAround()
-                ? isAfterStart || isBeforeEnd
-                : isAfterStart && isBeforeEnd;
+    @VisibleForTesting
+    boolean contains(Token token, boolean isStart) {
+        if (isEmpty()) {
+            return false;
+        }
+        Token minToken = factory.minToken();
+        if (end.equals(minToken)) {
+            if (start.equals(minToken)) { // ]min, min] = full ring, contains everything
+                return true;
+            } else if (token.equals(minToken)) {
+                return !isStart;
+            } else {
+                return isStart ? token.compareTo(start) >= 0 : token.compareTo(start) > 0;
+            }
+        } else {
+            boolean isAfterStart = isStart ? token.compareTo(start) >= 0 : token.compareTo(start) > 0;
+            boolean isBeforeEnd = isStart ? token.compareTo(end) < 0 : token.compareTo(end) <= 0;
+            return isWrappedAround()
+                    ? isAfterStart || isBeforeEnd
+                    : isAfterStart && isBeforeEnd;
+        }
```

Comment:
```java
if (token.equals(minToken)) {
        return isStart
                ? start.equals(minToken)
```
Not necessarily, for example this fails with the test "]2,1] contains the start of ]min,5]".

I need to merge this and merge to the downstream branches, so I'm going to proceed with the current version.